src/one/nio/net/native/socket.c

```diff
@@ -88,10 +88,19 @@ static int sockaddr_from_java(JNIEnv* env, jbyteArray address, jint port, struct
 
 static int is_udp_socket(int fd) {
     int type = 0;
-    int length = sizeof(type);
+    socklen_t length = sizeof(type);
     return getsockopt(fd, SOL_SOCKET, SO_TYPE, &type, &length) == 0 && type == SOCK_DGRAM;
 }
 
+static int get_int_socket_opt(JNIEnv* env, int fd, int level, int optname){
+    int optval = 0;
+    socklen_t length = sizeof(optval);
+    if(getsockopt(fd, level, optname, &optval, &length) == 0){
+        return optval;
+    }
+    throw_io_exception(env);
```

Comment:
Throwing `IOException` from methods not declared to throw any unchecked exception.

Reply:
I guess this is equivalent to JavaInternals#uncheckedThrow.

I wouldn't like to force users to check for exception while calling e.g. getRecvBuffer. I also wouldn't like to return 0 or false if getsockopt returned error. I think it makes sense to raise an exception.

What would you suggest?

Reply:
JavaInternals.uncheckedThrow is not encouraged either.

I don't expect getsockopt to fail unless it is called on a closed socket. For this reason I'm also against declaring `IOException` in method signatures. So, the question is what to return when calling getters on a closed socket. `getLocalAddress()`/`getRemoteAddress()` already return null in this case, and I think this is acceptable. So, I don't mind returning 0/false for other options, too. `IllegalStateException` also seems a reasonable outcome - indeed it means an operation on a socket in a wrong state. And, unlike `IOException`, it does not fool user by throwing unspecified exception.

Reply:
Agree fully. Also thinking now. Failing with ISE doesn't add much added value as the code is going to fail on read/write anyway.

I've changed the code to be consistent with returning 0/false instead of throwing exception.